lib/Sema/CSSimplify.cpp

```diff
@@ -5514,6 +5519,26 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
       return SolutionKind::Solved;
     };
 
+  auto tryMatchRootAndValueFromFunctionType =
+      [&](FunctionType *fnTy) -> SolutionKind {
```

Comment:
Could you please reuse logic for matching for root/value types between here and `tryMatchRootAndValueFromKeyPathType`?

Reply:
Even better would be to make it so `tryMatchRootAndValueFromKeyPathType` accepts a `Type` and then figures out whether it's a function or bound generic, or something else. That way we wouldn't need special cases for function type vs. bound generic in multiple places.

Reply:
Good idea, done!